chapters/3/modul-kompilujacy.tex

```diff
@@ -0,0 +1,75 @@
+
+\subsection{Wykorzystane pakiety}
+Wszystkie pakiety konieczne do prawidłowej pracy kontenera zostały pozyskane z repozytorium Ubuntu 20.04 przy pomocy narzędzia apt-get. Obejmowały kompilator mingw, narzędzie Wine pozwalające na uruchamianie aplikacji przeznaczonych na system operacyjny Windows oraz wirtualny serwer Xvfb zapewniający wirtualne środowisko graficzne dla Wine.
+
+\subsection{Wymiana plików między kontenerem, a systemem hosta}
+Narzędzie Docker zapewnia dwie podstawowe metody wymiany plików między kontenerem, a systemem hosta - Docker Volume\cite{dockerVolume} oraz Bind mounts\cite{dockerBindMounts}. Docker Volume jest mechanizmem zarządzanym przez system Docker, odpowiada on za tworzenie, zarządzanie i usuwanie danych. Możliwe jest współdzielenie jednego wolumenu przez liczne kontenery. Bind mounts to prostsze rozwiązanie polegające na bezpośrednim połączeniu wybranego katalogu w systemie hosta z katalogiem kontenera. Zapewnia to większą elastyczność, ponieważ nie wymaga tworzenia osobnego wolumenu, a jedynie powiązania ścieżki bezwzględnej na systemie hosta, ze ścieżką wewnątrz kontenera, jednak tym samym zmniejsza to izolację kontenera od systemu. W naszym rozwiązaniu zostały wykorzystane bind mounts, ponieważ każde sprawdzane rozwiązanie otrzymuje swój osobny folder powiązany z jego id, który wiążemy z osobnym, nowo utworzonym kontenerem. Eliminuje to potrzebę zarządzania licznymi wolumenami lub dzielenia wolumenu między kilka kontenerów co z kolei zmniejszyłoby izolację między zadaniami. Aby zaimplementowany kontener poprawnie skompilował i uruchomił program, struktura powiązanego folderu powinna wyglądać w następujący sposób:
+\begin{verbatim}
+    /folder powiązany
+    |-- /source
+    |   |-- main.cpp
+    |-- /input
+    |   |-- 1.in
+    |   |-- 2.in
+    |   |-- n.in
+    |-- /output
+        |-- 1.out
+        |-- 2.out
+        |-- n.out
+\end{verbatim}
+W katalogu /source znajdują się pliki zawierające kod źródłowy c++. Wejście, które chcemy przekazać przy uruchamianiu programu znajduje się w plikach *.in w folderze /input. W trakcie działania programu tworzone są pliki wynikowe z rozszerzeniem *.out w folderze /output. 
+
+\subsection{Skrypt uruchamiany przy starcie kontenera}
+Skrypt entrypoint to skrypt wskazany w pliku Dockerfile, który uruchamia się po starcie kontenera. Został napisany w Bashu i realizuje funkcjonalność modułu kompilującego, kompiluje pliki *.cpp znajdujące się w katalogu /source powiązanego folderu za pomocą mingw, a następnie uruchamia skompilowany program dla każdego z plików wejścia *.in umieszczonych w folderze /in. Wynik każdego uruchomienia przekierowywany jest do  pliku wynikowego *.out w folderze /out katalogu współdzielonego. Każde uruchomienie dodatkowo mierzy czas wykonania przy pomocy komendy date.
```

Comment:
>Skrypt entrypoint to skrypt wskazany w pliku Dockerfile

- Jakiś polski odpowiednik na entrypoint.
- Powtórzenie słowa "skrypt", ale to mniej istotne